doc/specifications/proposed/replicator/architecture.rst

```diff
@@ -0,0 +1,142 @@
+==========
+Replicator
+==========
+
+This document describes a common synchronization mechanism between two systems,
+which can cope with write failures, network downtimes and other types of
+partitions. The approach only works with a single write master (Single Source
+of Truth) and any number of read slaves.
+
+In our case we want to use this approach to synchronize the content database
+with a search index.
+
+General Workings
+================
+
+The general scheme for synchronizations is fairly simple:
+
+.. image:: synchronize.png
+
+It generally can be described as a three step process:
+
+1) Fetch last revision from target
+
+   The target is asked for the last revision the target knows. This is the
+   revision of the last successful write operation. For the first request the
+   target may return something like ``null``.
+
+2) Fetch all changes since that last revision from source
+
+   The source (write-master) is asked for all updates since the last revision.
+   A limit, to limit the transfer size, may also be applied. The results
+   **must** be sorted by their revision.
+
+3) Commit all updates on target
+
+   All fetched updates are committed on the target. The target **must not**
+   store a revision if the corresponding update did not succeed.
+
+   This is usually possible to implement using transactions. Another way is to
+   store the revision directly in the received update (document). This way, if
+   the document update is lost, the revision will also be "lost".
+
+Revisions
+---------
+
+Revisions **must** always be **globally strictly monotonically** increasing.
+
+If you have only one write-master time (including milliseconds) might work,
+unless there are multiple commits in the same millisecond or some NTP daemon
+stops time or makes it go backwards. Time will never be synchronous across
+multiple systems, so **do not** rely on that.
+
+A simple auto-increment in the write-master might be the simplest solution. You
+should take care for possible overflows, though. The revision might increase
+quite fast in scenarios with a lot of updates.
+
+Idempotent Updates
+------------------
+
+In some cases it is helpful if all updates are idempotent, thus applying the
+same update multiple times does no harm.
+
+In this case the last revision on the target system can be reset without
+resetting the full storage. This can be helpful for graceful recovery of the
+target system.
+
+Compaction
+----------
+
+One problem with this approach is that if the replication target had gone away
+every revision of every document will be re-transmitted in order. This can lead
+to quite long initializations.
+
+An option commonly used is to compact the database, thus removing deleted
+objects and older revisions so that they won't be re-transmitted. One must pay
+attention to not affect data integrity when doing this.
```

Comment:
The big problem with legacy content engine is that a lot of content operations are not atomic.
Which makes compaction an hazard.
E.g. add content, edit content, delete content => compaction means the 3 events could be removed from the queue.
But what if any of the slaves has already synchronized the 1st event?